packages/edit-post/src/components/manage-blocks-modal/manager.js

```diff
@@ -0,0 +1,193 @@
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Filters an item list given a search term.
+ *
+ * @param {Object[]} items      Set of inserter items to search within.
+ * @param {Object[]} categories Set of categories to search within.
+ * @param {string}   searchTerm Search term.
+ *
+ * @return {Object[]} Filtered item list.
+ */
+export const searchItems = ( items, categories, searchTerm ) => {
+	const normalizedSearchTerm = normalizeTerm( searchTerm );
+	const matchSearch = ( string ) => normalizeTerm( string ).includes( normalizedSearchTerm );
+
+	return items.filter( ( item ) => {
+		const itemCategory = find( categories, { slug: item.category } );
+
+		return (
+			matchSearch( item.title ) ||
+			some( item.keywords, matchSearch ) ||
+			( itemCategory && matchSearch( itemCategory.title ) )
+		);
+	} );
+};
+
+/**
+ * Converts the search term into a normalized term.
+ *
+ * @param {string} term The search term to normalize.
+ *
+ * @return {string} The normalized search term.
+ */
+export const normalizeTerm = ( term ) => {
+	// Disregard diacritics.
+	//  Input: "média"
+	term = deburr( term );
+
+	// Accommodate leading slash, matching autocomplete expectations.
+	//  Input: "/media"
+	term = term.replace( /^\//, '' );
+
+	// Lowercase.
+	//  Input: "MEDIA"
+	term = term.toLowerCase();
+
+	// Strip leading and trailing whitespace.
+	//  Input: " media "
+	term = term.trim();
+
+	return term;
+};
+
+function BlockManager( {
+	instanceId,
+	categories,
+	blockTypes,
+	disabledBlockTypes,
+	enableBlockTypes,
+	disableBlockTypes,
+	search,
+	setState,
+} ) {
+	const blockItems = blockTypes.map( ( blockType ) => ( {
+		id: blockType.name,
+		icon: blockType.icon,
+		title: blockType.title,
+		category: blockType.category,
+		hasChildBlocksWithInserterSupport: false,
+		className: classnames( 'edit-post-manage-blocks-modal__block-type', {
+			'is-inactive': includes( disabledBlockTypes, blockType.name ),
+		} ),
+	} ) );
+
+	const filteredBlockItems = search ?
+		searchItems( blockItems, categories, search ) :
+		blockItems;
+
+	// Disable reason (no-autofocus): The block manager is a modal display, not
+	// one which is always visible, and one which already incurs this behavior
+	// of autoFocus via Popover's focusOnMount.
+
+	/* eslint-disable jsx-a11y/no-autofocus */
+	return (
+		<div className="edit-post-manage-blocks-modal__content">
+			<label
+				htmlFor={ `edit-post-manage-blocks-modal__search-${ instanceId }` }
+				className="screen-reader-text"
+			>
+				{ __( 'Search for a block' ) }
+			</label>
+			<input
+				id={ `edit-post-manage-blocks-modal__search-${ instanceId }` }
+				type="search"
+				placeholder={ __( 'Search for a block' ) }
+				className="edit-post-manage-blocks-modal__search"
+				value={ search }
+				onInput={
+					( event ) => setState( { search: event.target.value } )
+				}
+				autoFocus
+			/>
+			<div className="edit-post-manage-blocks-modal__results">
```

Comment:
I'm not an accesibility expert but it seems like each panel should be somehow annotated with a region similar to the inserter to announce what is this list of blocks about:

https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/inserter/menu.js#L268-L274